### ISSUE TITLE
[BUG] Series Groupby dropping values

### ISSUE THREAD

**Issue by esnvidia** [rapidsai/cudf#11290]
**Describe the bug**
Series Groupby drops values and only returns index as a DataFrame.

**Steps/Code to reproduce bug**
```
import cudf
import cupy as cp

cp.random.seed(1234)

series = cudf.Series(data=cp.random.randint(0, 500, size=10000, dtype=cp.int16), name='count')

series.groupby('count').count()

series.value_counts()

series.groupby(series.values).count()

series.groupby(series.unique()).count()

series.groupby(series.values).count().sort_index()

series.value_counts().sort_index()

cp.all( (series.value_counts().sort_index() == series.groupby(series.values).count().sort_index()).values)

```

Above can be run in an ipython console/jupyter notebook.

**Expected behavior**

series.groupby('name').count() == series.value_counts()
or
series.groupby(series.values).count() == series.value_counts()
or
series.groupby(series.unique()).count() == series.value_counts()

**Environment overview (please complete the following information)**
Bare Metal, Rapids 22.06 r495 CUDA 11.5

**Comment by shwina** [comment]
Hi @esnvidia. Addressing each of your expected behaviours below:

> series.groupby('name').count() == series.value_counts()

it looks like you can't group by the name of a `Series` in Pandas. Taking your example:

```python
>>> series = pd.Series(data=np.random.randint(0, 500, size=10000, dtype=cp.int16), name='count')
>>> series.groupby('count').count()
# this gives a KeyError in Pandas
```

Instead, you want:

```python
series.groupby(series).count()
```

We should probably do a better job and raise an informative error message here, rather than forging ahead with the groupby.

> series.groupby('name').count() == series.value_counts()

`groupby` in cuDF does not guarantee the ordering of the result by default, so when comparing groupby results you coulld  sort by the index:


```python
>>> series = cudf.Series(data=cp.random.randint(0, 500, size=10000, dtype=cp.int16), name='count')
>>> (series.groupby(series).count().sort_index() == series.value_counts().sort_index()).all()
>>> True
```

> series.groupby(series.unique()).count() == series.value_counts()

Wouldn't this generally be `False`? For example:

```python
>>> s = pd.Series([1, 1, 2, 3])
>>> s.groupby(pd.Series(s.unique())).count()
Out[75]:
1.0    1
2.0    1
3.0    1
dtype: int64
>>> s.value_counts()
1    2
2    1
3    1
dtype: int64
```

**Comment by esnvidia** [comment]
> it looks like you can't group by the name of a Series in Pandas.
Yes, just seems _odd_. Also seems like there's a silent bug there. Am aware of this `series.groupby(series).count()`

> ... so when comparing groupby results you coulld sort by the index:
yes, I do that in the last line in my example w/ `cp.all(...)`

> Wouldn't this generally be False?
` s.groupby(s.unique()).count()`

I understand the output is different, the question is _why_ it is False.
Perhaps there's a logic error, but I would have imagined that if I said "group the series S by these Z values" that it would be the same whether the Z values were unique or not - since the number of groups is the cardinality of the unique set.

In other words, seems like the values passed to the `by=` parameter provides the set of groups so that
```
# very naively...
by=[1,1,2,3] 
groups = {'unknown': []}
for value in by:
    if value not in groups:
        groups[value] = []

# now that we have the groups, go add the series S values
for value in series:
    if value in groups:
        groups[value].append(value)
    else:
        groups['unknown'].append(value)

# count
counts = {}
for group in groups:
    counts[group] = len(group)
```

above will work whether `by=[1,1,2,3] ` or `by=[1,2,3]`

**Comment by shwina** [comment]
Thanks - I think I understand the source of the confusion:

Here's a clarifying example:

```python
>>> s = cudf.Series([1, 2, 3, 4])
>>> z = cudf.Series([5, 6, 5, 6])
>>> s.groupby(z).sum()
Out[41]:
5    4
6    6
dtype: int64
```

The Series `s` is being grouped by the values in `z`. There are two groups, corresponding to the 2 unique values in `z` (`5` and `6`). The grouping is very much dependent on the number and location of those unique values. The locations of the value `5` are  at indexes `0` and `2`, so the group sum is `s[0] + s[2] == 4`. The locations of the value `6` are at indexes `1` and `3`, so the group sum is `s[1] + s[3] == 6`.

In general, the number of elements in `z` should be equal to the number of values in `s`. When there is a mismatch, I believe Pandas will align the indexes of `s` and `z` before doing the groupby. Thus (in Pandas):

```python
>>> s = pd.Series([1, 2, 3, 4])
>>> z = np.array([5, 5, 6])
>>>  s.groupby(z).sum()  # this won't work because `s` and `z` differ in length
>>> s.groupby(pd.Series(z)).sum()  # this will work by aligning the indexes first
```

**Comment by vyasr** [comment]
@esnvidia did @shwina's last comment help address your confusion? Is there anything additional that needs to happen here?

**Comment by vyasr** [comment]
I'm going to close this as solved - the confusion was around the fact that we need a group per row, not just the set of unique values that exist - but please feel free to reopen if needed.
